plugins/VersionUpgrade/VersionUpgrade31to32/VersionUpgrade31to32.py

```diff
@@ -0,0 +1,212 @@
+
+                        # For single extruder machine we need only first extruder
+                        if len(machine_extruders) != 0:
+                            self._updateSingleExtruderDefinitionFile(machine_extruders, machine_nozzle_size)
+                            parser.remove_option("values", "machine_nozzle_size")
+
+        # Update version numbers
+        parser["metadata"]["setting_version"] = "5"
+
+    def _getSingleExtrusionMachineExtruders(self, definition_name):
+        machine_instances_dir = Resources.getPath(CuraApplication.ResourceTypes.MachineStack)
+        machine_instance_id = None
+        extruder_instances = []
+
+        # Find machine instances
+        for item in os.listdir(machine_instances_dir):
+            file_path = os.path.join(machine_instances_dir, item)
+            if not os.path.isfile(file_path):
+                continue
+
+            parser = configparser.ConfigParser(interpolation=None)
+            try:
+                parser.read([file_path])
+            except:
+                # skip, it is not a valid stack file
+                continue
+
+            if not parser.has_option("metadata", "type"):
+                continue
+            if "machine" != parser["metadata"]["type"]:
+                continue
+
+            if not parser.has_option("general", "id"):
+                continue
+
+            machine_id = parser["general"]["id"]
+            if machine_id + "_settings" != definition_name:
+                continue
+            else:
+                machine_instance_id = machine_id
+                break
+
+        if machine_instance_id is not None:
+            extruders_instances_dir = Resources.getPath(CuraApplication.ResourceTypes.ExtruderStack)
+
+            # Find all custom extruders for found machines
+            for item in os.listdir(extruders_instances_dir):
+                file_path = os.path.join(extruders_instances_dir, item)
+                if not os.path.isfile(file_path):
+                    continue
+
+                parser = configparser.ConfigParser(interpolation=None)
+                try:
+                    parser.read([file_path])
+                except:
+                    # skip, it is not a valid stack file
+                    continue
+
+                if not parser.has_option("metadata", "type"):
+                    continue
+                if "extruder_train" != parser["metadata"]["type"]:
+                    continue
+
+                if not parser.has_option("metadata", "machine"):
+                    continue
+                if not parser.has_option("metadata", "position"):
+                    continue
+
+                if machine_instance_id != parser["metadata"]["machine"]:
+                    continue
+
+                extruder_instances.append(parser)
+
+        return extruder_instances
+
+    # Find extruder definition at index 0 and update its values
```

Comment:
This is not correct. Index 0 is user_changes. Settings such as nozzle size and material diameter should be save in the definition_changes container at index 5